src/array_indexing.rs

```diff
@@ -24,28 +26,122 @@ declare_lint! {
     "out of bound constant indexing"
 }
 
+/// **What it does:** Check for usage of indexing or slicing.
+///
+/// **Why is this bad?** Usually, this can be safely allowed. However,
+/// in some domains such as kernel development, a panic can cause the
+/// whole operating system to crash.
+///
+/// **Known problems:** Hopefully none.
+///
+/// **Example:**
+///
+/// ```
+/// ...
+/// x[2];
+/// &x[0..2];
+/// ```
+declare_lint! {
+    pub INDEXING_SLICING,
+    Allow,
+    "indexing/slicing usage"
+}
+
 #[derive(Copy,Clone)]
 pub struct ArrayIndexing;
 
 impl LintPass for ArrayIndexing {
     fn get_lints(&self) -> LintArray {
-        lint_array!(OUT_OF_BOUNDS_INDEXING)
+        lint_array!(INDEXING_SLICING, OUT_OF_BOUNDS_INDEXING)
     }
 }
 
 impl LateLintPass for ArrayIndexing {
     fn check_expr(&mut self, cx: &LateContext, e: &Expr) {
         if let ExprIndex(ref array, ref index) = e.node {
-            let ty = cx.tcx.expr_ty(array);
+            // Ranges without begin and end are always valid
+            if let Some(range) = utils::unsugar_range(index) {
+                if range.start.is_none() && range.end.is_none() {
+                    return;
+                }
+            }
 
+            // Array with known size can be checked statically
+            let ty = cx.tcx.expr_ty(array);
             if let TyArray(_, size) = ty.sty {
-                let index = eval_const_expr_partial(cx.tcx, &index, ExprTypeChecked, None);
-                if let Ok(ConstVal::Uint(index)) = index {
-                    if size as u64 <= index {
-                        span_lint(cx, OUT_OF_BOUNDS_INDEXING, e.span, "const index-expr is out of bounds");
+                let size = size as u64;
+
+                // Index is a constant uint
+                let const_index = eval_const_expr_partial(cx.tcx, &index, ExprTypeChecked, None);
+                if let Ok(ConstVal::Uint(const_index)) = const_index {
+                    if size <= const_index {
+                        utils::span_lint(cx, OUT_OF_BOUNDS_INDEXING, e.span, "const index is out of bounds");
+                        utils::span_lint(cx, INDEXING_SLICING, e.span, "indexing/slicing may panic");
+                    } else {
+                        // Index is within bounds
+                        return;
+                    }
+                }
+
+                // Index is a constant range
+                if let Some(range) = utils::unsugar_range(index) {
+                    let start: Option<Result<ConstVal, ()>> = range.start.map(|start|
+                        eval_const_expr_partial(cx.tcx, start, ExprTypeChecked, None)
+                        .map_err(|_| ()));
```

Comment:
You should be able to write `range.start.map_or(None, |start| Some(eval_const_expr_partial(…)))` or something